### PR TITLE
Manage Linear feature projects and specification documents

### PR DESCRIPTION
**Pull request by howarewoo**
## Goal

Add idempotent Linear feature-project and specification-document operations with lifecycle, ownership, concurrency, and read-back guarantees.

## Summary

- Add Linear project and document GraphQL operations plus the `linear.sh` command surface for preflight, feature resolution/creation/transition, and spec read/write.
- Resolve explicit UUIDs and URLs exactly; otherwise discover only canonical repository-marked projects in operation-valid lifecycle states.
- Recover safely from unknown and partial mutation outcomes through discovery and byte-exact read-back; emit content-free receipts with attempted, observed, returned, verified, and pending fields.
- Enforce canonical managed metadata, immutable ownership, optimistic revisions, valid forward lifecycle transitions, explicit abandon, pagination progress, and terminal-state protection.
- Add offline fixtures and tests for create/resume, duplicates, malformed metadata, pagination, read-back mismatches, content privacy, lifecycle transitions, and partial responses.
- Mark Increment 3 complete in the implementation plan.

### Automated

- [x] `bash skills/woostack-init/scripts/tests/test-linear-resources.sh` — 103 passed, 0 failed.
- [x] `bash skills/woostack-init/scripts/tests/test-linear-transport.sh` — 162 passed, 0 failed.
- [x] `bash skills/woostack-init/scripts/tests/test-linear-metadata.sh` — 66 passed, 0 failed.
- [x] `bash -n` passed for changed shell scripts.
- [x] `shellcheck` passed for changed shell scripts.
- [x] Task-level spec-compliance and code-quality reviews approved.

### Manual

- [ ] Confirm PR #499 is stacked on PR #498.
- [ ] Confirm receipts contain identifiers and verification facts but no specification content, credentials, or local paths.
- [ ] Confirm no increment issue or relation operations are included in this increment.

Spec: .woostack/specs/2026-07-11-linear-artifact-backend.md

### CODE REVIEW

**Comment by howarewoo** [comment]
* **#506** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/506?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#505** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#504** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/504?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#503** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/503?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#502** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/502?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#501** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/501?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#500** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/500?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#499** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/499?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/howarewoo/woostack/499?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#498** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#497** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/497?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#495** <a href="https://app.graphite.com/github/pr/howarewoo/woostack/495?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: skills/woostack-init/scripts/artifacts/linear.sh [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+  fi
+}
+
+command_preflight() {
+  local workspace='' team='' project_statuses='' issue_states='' response tmp
+  while [[ $# -gt 0 ]]; do case "$1" in
+    --workspace) workspace="$2"; shift 2;; --team) team="$2"; shift 2;;
+    --project-statuses) project_statuses="$2"; shift 2;; --issue-states) issue_states="$2"; shift 2;; *) usage;; esac; done
+  [[ -n "$workspace" && -n "$team" && -n "$project_statuses" && -n "$issue_states" ]] || usage
+  jq -e --argjson required "$REQUIRED_STATUSES" 'type=="object" and ((keys|sort)==($required|sort))' >/dev/null 2>&1 <<<"$project_statuses" || fail "all project statuses, including abandoned, are required"
```
Comment (howarewoo) [inline skills/woostack-init/scripts/artifacts/linear.sh]: **Preflight accepts aliased lifecycle mappings**

These checks validate only the semantic keys. If two semantics name the same native status or state, `linear-preflight.sh` resolves both to the same ID and reports success; the returned project map is then rejected by `validate_status_map`, while aliased issue states remain ambiguous for later lifecycle operations. Thus authenticated preflight can approve a configuration that its own command surface cannot safely use.

Fix: Before issuing the preflight query, require every configured lifecycle value to be a nonblank string and require the values in both the project-status and issue-state maps to be unique; add duplicate-value cases to the focused preflight tests.

<sub>MEDIUM · NIT</sub>

File: skills/woostack-init/scripts/artifacts/linear.sh [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+  while [[ $# -gt 0 ]]; do case "$1" in
+    --repository) repository="$2"; shift 2;; --title) title="$2"; shift 2;; --team-id) team_id="$2"; shift 2;;
+    --status-map) status_map="$2"; shift 2;; --spec-file) spec_file="$2"; shift 2;; *) usage;; esac; done
+  require_repository "$repository"; [[ -n "$title" ]] || usage; require_uuid "$team_id" "team id"; validate_status_map "$status_map"; [[ -r "$spec_file" ]] || fail "spec file is unreadable"
+  projects="$(project_list)" || fail "project discovery failed"
+  set +e; project="$(resolve_from_list "$projects" "$repository" "$status_map" "$ACTIVE_STATUSES" '' 2>/dev/null)"; resolve_rc=$?; set -e
+  case "$resolve_rc" in
+    0) ;;
+    3)
+      terminal="$(managed_projects "$projects" "$repository" "$status_map" '["done","abandoned"]')"
```
Comment (howarewoo) [inline skills/woostack-init/scripts/artifacts/linear.sh]: **Terminal projects prevent every future feature**

When no active project exists, this collects every unarchived managed project in `done` or `abandoned` and aborts if even one exists. Completed and abandoned projects are deliberately preserved, so after the repository finishes or abandons its first Linear feature, every later `feature-create` fails even though none of those terminal projects is eligible for resume.

Fix: Ignore terminal projects when deciding whether a new draft project may be created; retain the active-project ambiguity check, and replace the regression test that currently enshrines the permanent terminal-state block with coverage that creates the next feature alongside preserved terminal history.

<sub>MEDIUM · NIT</sub>

File: skills/woostack-init/scripts/tests/test-linear-resources.sh [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+run_capture feature-create --repository acme/widgets --title 'Feature Alpha' --team-id "$team_id" --status-map "$status_map" --spec-file "$work/spec.md"
+assert_exit 0 "$RC" "project and spec create succeeds"
+assert_eq "$(jq -r '.verified' <<<"$OUTPUT")" "true" "create receipt is verified"
+for field in attempted observed returned verified pending; do
+  if jq -e --arg field "$field" 'has($field)' <<<"$OUTPUT" >/dev/null; then pass; else fail "create receipt includes $field"; fi
+done
+assert_eq "$(grep -c $'mutation\tproject-create' "$work/calls")" "1" "project mutation occurs once"
+assert_eq "$(grep -c $'mutation\tdocument-create' "$work/calls")" "1" "document mutation occurs once"
+
+# Unknown mutation outcome is discovered and resumed without repeating the mutation.
```
Comment (howarewoo) [inline skills/woostack-init/scripts/tests/test-linear-resources.sh:151]: **Recovery failures are not exercised**

The unknown-outcome cases only queue successful follow-up discovery. None exercises the explicit branches where the required project or document read-back itself fails, so regressions in the unverified receipt shape, pending reason, nonzero exit, or no-retry guarantee would pass this suite even though failed read-back is the safety-critical ambiguous-outcome path.

Fix: Add focused cases that make the post-mutation project-list or document-list request fail for create, spec-write, and feature-transition, then assert a nonzero exit, verified=false, the command-specific read-back pending reason, a content-free receipt, and exactly one mutation call.

<sub>MEDIUM · NIT</sub>

File: skills/woostack-init/scripts/tests/test-linear-resources.sh [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+make_document_list "$work/zero-list.json" "$work/zero-newline.md" '2026-07-12T10:06:00.000Z'
+reset_fake; queue document-list document-list-one.json 1; queue document-update document-update-partial.json 1; cp "$work/zero-list.json" "$work/responses/document-list.2.json"
+run_capture spec-write --project "$project_id" --repository acme/widgets --content-file "$work/zero-newline.md" --expected-revision "$revision"
+assert_exit 0 "$RC" "zero-terminal-newline content verifies exactly"
+printf '\n' >>"$work/zero-newline.md"
+reset_fake; queue document-list document-list-one.json 1; queue document-update document-update-partial.json 1; cp "$work/zero-list.json" "$work/responses/document-list.2.json"
+run_capture spec-write --project "$project_id" --repository acme/widgets --content-file "$work/zero-newline.md" --expected-revision "$revision"
+assert_exit 1 "$RC" "one versus zero terminal newline mismatch fails"
+printf '\n' >>"$work/zero-newline.md"
+make_document_list "$work/one-list.json" "$work/revised.md" '2026-07-12T10:07:00.000Z'
```
Comment (howarewoo) [inline skills/woostack-init/scripts/tests/test-linear-resources.sh]: **Newline regression has a second mismatch**

The multiple-terminal-newline case compares zero-newline.md, whose body says `Zero newline.`, with one-list.json generated from revised.md, whose body says `Revised body.`. The command therefore fails even if it incorrectly normalizes multiple terminal newlines to one, so this assertion cannot detect the boundary regression named by the test.

Fix: Generate the read-back fixture from a copy of zero-newline.md that differs only by having exactly one terminal newline, leaving every other byte identical to the two-newline input.

<sub>MEDIUM · NIT</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

The full review found four non-blocking correctness and coverage nits in Linear feature/project lifecycle handling.

---
**Status: APPROVED** — No blocking findings, 4 nit(s). See inline comments.
*Audited by woostack-review · Host: omp · Provider: openai-codex · Model: gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks APPROVE on your own PR. Status line above carries the actual verdict._

File: skills/woostack-init/scripts/artifacts/linear.sh (hunk 1 above)
Comment (howarewoo) [reply skills/woostack-init/scripts/artifacts/linear.sh]: Fixed in 3825735. Preflight now requires nonblank, unique configured values for both project statuses and issue states before making any request, with focused alias tests.

File: skills/woostack-init/scripts/artifacts/linear.sh (hunk 2 above)
Comment (howarewoo) [reply skills/woostack-init/scripts/artifacts/linear.sh]: Fixed in 3825735. Terminal managed projects remain preserved history but no longer block creation of the next draft feature; the regression now verifies one new project mutation beside terminal history.

File: skills/woostack-init/scripts/tests/test-linear-resources.sh (hunk 3 above)
Comment (howarewoo) [reply skills/woostack-init/scripts/tests/test-linear-resources.sh:151]: Fixed in 3825735. Focused tests now cover failed read-back after project create, document create, spec update, and project transition, asserting unverified content-free receipts, precise pending reasons, nonzero exits, and exactly one mutation.

File: skills/woostack-init/scripts/tests/test-linear-resources.sh (hunk 4 above)
Comment (howarewoo) [reply skills/woostack-init/scripts/tests/test-linear-resources.sh]: Fixed in 3825735. The one-newline read-back fixture is now derived from the same two-newline input by removing exactly one byte, so content differs only at the terminal-newline boundary.

File: skills/woostack-init/scripts/tests/test-linear-resources.sh [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+    --variables) variables="$2"; shift 2 ;;
+    *) exit 2 ;;
+  esac
+done
+name="$(basename "$document" .graphql)"
+count_file="$LINEAR_FAKE_STATE/$name.count"
+count=0; [ ! -f "$count_file" ] || count="$(cat "$count_file")"
+count=$((count + 1)); printf '%s' "$count" >"$count_file"
+printf '%s\t%s\t%s\n' "$operation" "$name" "$variables" >>"$LINEAR_FAKE_STATE/calls"
+response="$LINEAR_FAKE_RESPONSES/$name.$count.json"
```
Comment (howarewoo) [inline skills/woostack-init/scripts/tests/test-linear-resources.sh:29]: **Transport fake ignores every variables payload**

The fake chooses each canned response solely from the operation name and call count, while the test body never inspects the logged `variables` field. As a result, the create, update, transition, and pagination cases still pass if `linear.sh` sends the wrong team/status/ownership/content or fails to forward the page cursor, because the fake returns the expected read-back independently of the request payload.

Fix: Assert the parsed GraphQL variables for each operation's key contract, including canonical project metadata and team/status IDs, document project/content, transition status ID, and the `after` cursor on second-page queries; alternatively make the fake select or reject responses based on those variables.

<sub>MEDIUM · NIT</sub>

**Review by howarewoo — COMMENTED** [review]
## AI Deep Code Review Summary

The second full review found one remaining non-blocking test-contract gap.

---
**Status: APPROVED** — No blocking findings, 1 nit(s). See inline comments.
*Audited by woostack-review · Host: omp · Provider: openai-codex · Model: gpt-5.6-sol*



_Review event downgraded to COMMENT — GitHub blocks APPROVE on your own PR. Status line above carries the actual verdict._

File: skills/woostack-init/scripts/tests/test-linear-resources.sh (hunk 5 above)
Comment (howarewoo) [reply skills/woostack-init/scripts/tests/test-linear-resources.sh:29]: Fixed in 03facf8. The focused suite now parses logged GraphQL variables and asserts project team/status/ownership, document project/content, spec update identity/content, transition status, and project/document pagination cursors.

**Comment by howarewoo** [comment]
### Merge activity

* **Jul 14, 7:27 PM UTC**: @howarewoo merged this pull request with [Graphite](https://app.graphite.com/github/pr/howarewoo/woostack/499).
